### PR TITLE
Implement `vec_arith()` and `vec_math()` for POSIXlt

### PR DESCRIPTION
**Pull request by DavisVaughan**
Closes https://github.com/r-lib/vctrs/issues/971

Required for https://github.com/tidyverse/lubridate/pull/871

I've also revised the tests a little bit so it is clearer to see that all of the methods are being tested symmetrically

### CODE REVIEW

**Comment by lionel-** [comment]
Should we wait until next version before adding arith methods for lubridate? We haven't updated the double dispatch mechanism yet.

Alternatively we could switch to the new mechanism now, I don't expect it should have much impact on revdeps.

**Comment by DavisVaughan** [comment]
If it was eventually implemented with the "legacy double dispatch" fallback mechanism then I didn't think it would matter too match to implement it in vctrs right now. lubridate could be updated later as required. The main reason I wanted to do it was to use it in slider, dplyr won't need it, so it can definitely wait.

**Comment by lionel-** [comment]
I meant that the semantics would change (inheritance).

**Comment by DavisVaughan** [comment]
Oh right. I hadn't thought about that too much for lubridate. I wasn't even planning on guarding against it.

I haven't seen many people subclassing Date, POSIXct, or lubridate Period objects. The only thing to potentially worry about with subclasses would be `<Period> * <numeric-subclass>` but that seems pretty rare too.

I guess the bigger problem would be that we currently export `vctrs::vec_arith.numeric` rather than individual methods for integer and double. This is https://github.com/r-lib/vctrs/issues/997. With the no-inheritance model I think we would switch to explicit integer/double methods, which would require changes in lubridate.

**Comment by DavisVaughan** [comment]
I'm happy to wait on lubridate vec-arith methods until the next release of vctrs.

If there are no objections to adding POSIXlt support, I do think we could still do that now.

**Comment by hadley** [comment]
I think we could change the dispatch mechanism if you wanted — would just need to submit PRs to the affected packages (https://github.com/search?q=org%3Acran+vec_arith&type=Code)

File: R/type-date-time.R [hunk 1]
```diff
@@ -427,6 +439,37 @@ vec_arith.difftime.Date <- function(op, x, y, ...) {
     stop_incompatible_op(op, x, y)
   )
 }
+#' @method vec_arith.POSIXct difftime
+#' @export
+vec_arith.POSIXct.difftime <- function(op, x, y, ...) {
+  y <- vec_cast(y, new_duration(units = "secs"))
+
+  switch(op,
+         `+` = vec_restore(vec_arith_base(op, x, y), x),
```
Comment (hadley) [inline R/type-date-time.R:448]: Indenting

Comment (DavisVaughan) [reply R/type-date-time.R:448]: I hate that RStudio does this automatically

**Pull request merged**